README.md
| 文档名称 | 《Vango-ERP 前端框架技术手册》 |
| -------- | ------------------------------ |
| 文档版本 | V1.0                           |
| 编辑人员 | 前端组                         |
| 更新历史 | 2021-06-07 文档创建            |

# 1. 前言

## 1.1 背景和阅读对象

jsp -> El-admin -> antdesign vue

## 1.2 现有框架问题

ant 社区不活跃、可参考文档较少，市场使用量少，组件存在 bug

权限部分实现待完善

路由未打开三级

组件封装度不够

...

## 1.3 技术选型

阐述 Vue3.0 和 element Plus 或其他技术选型理由，并附上相关技术点文档参考链接

如 渲染优化、打包体积启动快、vite 介绍等

```
**1.3.1** 项目技术选型
项目框架使用vue3
vue3中文文档:https://vue3js.cn/docs/zh/guide/migration/introduction.html

vue3相对于vue2 来说更小、更快、加强TypeScript支持、加强API设计一致性、提高自身可维护性、开放更多底层功能
压缩包体积将缩小到原来的将近一半，vue3将vue2的拦截监听使用的Object.defineProperty换成了ES6的新特性 Proxy(代理),proxy代理了对象,而不是代理了对		象的属性,这样可以将原本对对象属性的操作变为对整个对象的操作，颗粒度变大。javascript引擎在解析的时候希望对象的结构越稳定越好，如果对象一直在变，可优化性	 降低，proxy不需要对原始对象做太多操作。
Virtual DOM 重构
部分指令更改或者去除
Composition API的引入
hooks的使用
Teleport的引入
生命周期更改



```

## 1.4 参考 志芳

https://adminpro.iviewui.com/

https://preview.pro.ant.design/

https://el-admin.xin/
https://element-plus.gitee.io/#/zh-CN/component/infiniteScroll -element-plus 官网

https://vue3js.cn/docs/zh/ -vue3.0 中文文档

https://next.vuex.vuejs.org/ -vueX4.0 文档

https://cn.vitejs.dev/ -Vite 官方中文文档

http://www.axios-js.com/ -axios 中文文档

https://next.router.vuejs.org/zh/index.html -Vue Router

http://lesscss.cn/ -Less 中文网

https://www.sass.hk/ -Sass 中文网

https://www.stylus-lang.cn/ -Stylus 中文网

https://www.iconfont.cn/ -iconfont 图标库

https://antv.vision/zh -蚂蚁数据可视化

https://echarts.apache.org/zh/index.html -Echarts 5.0 可视化图表库

# 2. 全局说明

## 2.1 概述

介绍项目整体技术点，特性，服务行业对象等

```
**2.1.1** 项目架构特点和整体技术点
ERP 项目采用自行搭建的后台管理架构进行开发,架构上和大多数管理系统模板相似,以高效、便捷、灵活为搭建理念。
   1：灵活的路由设计、权限设计、人员角色管理
    以管理员账号身份进入角色管理，和菜单管理界面进行角色和权限的管理，用表格设计建立角色的树形结构体系和各个角色以及角色标识
    以表格设计建立菜单树形结构体系且可灵活关联上用户角色的权限。
   2：项目采用vue3框架
    vue3相对于vue2 来说更小、更快、加强TypeScript支持、加强API设计一致性、提高自身可维护性、开放更多底层功能
    diff算法的重构
    监听拦截的重构
    Virtual DOM 重构
    vue3 引入 Composition API
   3：项目采用Vite编译
   	之前项目通常采用webpack编译，Webpack Dev Server 在启动时，需要先 build 一遍，而 build 的过程是需要耗费很多时间的。
   	而 Vite 则完全不同，当我们执行 vite serve 时，内部直接启动了 Web Server，并不会先编译所有的代码文件。
	Webpack 这类工具的做法是将所有模块提前编译、打包进 bundle 里，换句话说，不管模块是否会被执行，都要被编译和打包到 bundle 里。随着项目越来越大打包后的     bundle 也越来越大，打包的速度自然也就越来越慢。
	Vite 利用现代浏览器原生支持 ESM 特性，省略了对模块的打包。对于需要编译的文件，Vite 采用的是另外一种模式：即时编译。也就是说，只有具体去请求某个文件时才     会编译这个文件。所以，这种「即时编译」的好处主要体现在：按需编译.

	Vite提供了一个子命令 optimize 这个命令的作用就是单独的去「优化依赖」

	HRM热更新方面
	Vite 只需要立即编译当前所修改的文件即可，所以响应速度非常快。
	Webpack 修改某个文件过后，会自动以这个文件为入口重写 build 一次，所有的涉及到的依赖也都会被加载一遍，所以反应速度会慢很多

**2.1.2** 服务行业对象
	该项目的架构的初衷就是为了服务大型多模块应用,因为此次立项的ERP系统涉及版块过多,本可以尽力往微前端方向靠近,使用模块化开发,后为了减少工作时间以及架构     设计时间,决定只用一个项目架构供应ERP的开发,但是在文件目录、vuex、权限和路由的设计上都力求保证模块化、清晰化。这样即可供应ERP开发且不会紊乱冗余。


```

​

## 2.2 目录

截图并说明项目详细目录结构，以及对后续业务和功能拆分拓展可能涉及到的目录说明

![](http://vshoptest.oss-cn-hangzhou.aliyuncs.com/0daf740bd89f55e9ad1fc0b9887159cd4d18922aca26b7e0d6099e95d6d0b03b-o.png)

| 文件夹/文件 名称         | 文件夹/文件 功能               | 备注 |
| ------------------------ | ------------------------------ | ---- |
| assets 文件夹            | 存放静态文件(img,js,css 等)    |      |
| mixins                   | 存放混入文件                   |      |
| router                   | 存放路由配置文件               |      |
| router/index.js          | 路由主配置文件                 |      |
| router/router.config.js  | 基础路由配置文件               |      |
| router/permission.js     | 权限校验文件                   |      |
| store                    | 存放 vuex 配置文件             |      |
| store/module             | vuex 模块存放                  |      |
| utils                    | 公用方法以及工具函数存放处     |      |
| utils/AbnormalMonitoring | 全局报错捕捉的配置文件存放     |      |
| utils/commonDirective    | 全局配置自定义指令             |      |
| utils/commonMixins       | 全局配置混入                   |      |
| utils/request.js         | 请求方法配置文件               |      |
| utils/elComponents.js    | element 组件按需引入           |      |
| utils/device.js          | 屏幕缩放比例监听和校正配置文件 |      |
| utils/vite.js            | 环境变量引入和配置文件         |      |
| views                    | 页面组件存放                   |      |
| views/Layout             | Layout 布局组件                |      |
| App.vue                  | 全局入口组件                   |      |
| main.js                  | 全局配置文件                   |      |
| .env.development         | 本地环境变量                   |      |
| .env.production          | 正式环境变量                   |      |
| vite.config.js           | vite 配置文件                  |      |

## 2.3 全局配置

vue 配置，vite 配置等

## 2.4 兼容性 朱涛

### 2.4.1 浏览器兼容性

对浏览器及版本号的兼容特性进行说明

### 2.4.2 分辨率适配

对可正常适配的最小分辨率以及最佳使用分辨率进行说明

简单介绍实现机制

```
在 utils 下新建了 device.js 文件用于监控页面大小缩,然后校正
```

## 2.5 环境变量 朱涛

列举一下环境涉及的全局变量和相关作用

**dev**

测试环境: _NODE_ENV = "development"_

构建后的基本公共路径: ITE_BASE = "/"\*

是否禁用 console: _VITE_DROP_CONSOLE = false_

是否禁用 debugger: _VITE_DROP_DEBUGGER = false_

api 地址: _VITE_API_URL = "/api"_

反向代理: _VITE_PROXY = [["/api", "http://xxx.xx.xxx"]]_

**test**

**prod**

测试环境: _NODE_ENV = "development"_

构建后的基本公共路径: _VITE_BASE = "/"_

是否禁用 console: _VITE_DROP_CONSOLE = false_

是否禁用 debugger: _VITE_DROP_DEBUGGER = false_

api 地址,完整服务端域名: _VITE_API_URL = "http://xx.xx.xx:8083"_

## 2.6 组件 朱涛 新国

对项目下回使用到的常用组件和样式选择进行说明（尽可能屏蔽不同开发人员组件使用的差异）

https://element-plus.gitee.io/#/zh-CN/component/installation

## 2.7 命令 朱涛

开发环境启动和关闭

_npm run dev_

_npm run build-dev_

部署环境（测试和生产）启动和关闭

_npm run build-prd_

# 3. 规范

## 3.1 命名规范 朱涛 王杰

代码规范

命名规范

## 3.2 注释规范 陈斌

格式化 插进配置

注释 文件注释，方法注释，代码注释

参考链接： https://blog.csdn.net/zxBlogs/article/details/106097389

```json
{
  "Print to vue temaplate": {
    "prefix": "vue",
    "body": [
      "/*",
      " *@Author: chenbin",
      " *@Date: ${CURRENT_YEAR}-${CURRENT_MONTH}-${CURRENT_DATE} ${CURRENT_HOUR}:${CURRENT_MINUTE}:${CURRENT_SECOND}",
      " *@Description: $0",
      "*/",
      "<template>",
      "  <div class=\"wrapper\">$0</div>",
      "</template>",
      "",
      "<script>",
      "export default {",
      "  name: '',",
      "  data () {",
      "    return {",
      "    }",
      "  },",
      "  methods: {},",
      "  mounted () {}",
      "}",
      "</script>",
      "<style lang=\"less\" scoped>",
      "</style>\r"
    ],
    "description": "A vue2 file template"
  }
}
```

文件注释规范：

文件注释用于告诉不熟悉这段代码的读者这个文件中包含哪些东西。 应该提供文件的大体内容, 它的作者, 依赖关系和兼容性信息。如下:
/\*\*

- @fileoverview Description of file, its uses and information
- about its dependencies.
- @author user@meizu.com (Firstname Lastname)
- Copyright 2015 Meizu Inc. All Rights Reserved.
  \*/

代码注释规范：

单行注释

- 注释单独一行的情况下，注释的//后面要跟一个空格
- 注释如果和代码同一行，代码分号结束后，要跟一个空格，注释的//后也要跟一个空格

多行注释

多行注释使用下面这种形式：

- 总是保持星号纵向对齐（结束符前留一个空格）
- 不要在开始符、结束符所在行写注释
- 尽量使用单行注释代替多行注释
- 注释函数时，推荐使用多行注释

函数注释

- 其间每一行都以 `*` 开头，且与第一行第一个 `*` 对齐

- 注释内容与 `*` 间留一个空格

- 必须包含标签注释。

/\*\*
_这是一个求和函数
_@param {Number} a 第一个数字
_@param {Number} b 第一个数字
_@return {Number} 返回两个数字之和
\*/
var sum = function(a,b){
return a + b;
}

# 4. 公共

## 4.1 公共组件 朱涛 新国

列举二次封装的组件和使用说明

##### 4.1.1 上传组件

##### 4.1.2 下载组件

##### 4.1.3 富文本组件

##### 4.1.4 日期选择器

##### 4.1.5 分页列表

##### 4.1.6 弹框（提示/警告/报错等）

## 4.2 共用工具

列举常用工具类

时间

字符串处理

小数处理

缓存封装

提示/弹框

防抖/限流

# 5. 架构设计

## 5.1 界面配置

菜单

在 views/layout 文件夹里自定义菜单样式和菜单渲染

图片

皮肤

现在 vuex 里配置了 theme 主题字段,在 App.vue 声明了一个基础的 --themeColor: #008cff;

可通过 vuex 全局更改这个样式

## 5.2 路由设计

1、安装

```
npm install vue-router@next
```

2、使用

```
// router/index.js
import { createRouter } from 'vue-router'
import routes from './routes'

import Login from "../pages/login"
import Home from "../pages/home"
import NotFound from "../pages/errorPage/404"
import Forbidden from "../pages/errorPage/403"
import Layout from "../pages/layout"

// 初始化路由
const routes = [
  {
    path: '/login',
    name: 'Login',
    component: Login
  }
]

// 准备动态加载的路由
export const DynamicRoutes = [
    {
        path:"",
        component:Layout,
        name:'container',
        redirect:"home",

CHANGELOG.md
# 1.0.0 (2021-08-26)

### 🌟 新功能
范围|描述|commitId
--|--|--
 - | 111222222 | [5398734](http://47.99.38.141:10080/web/vango-erpweb/commits/5398734)
 - | 11菜单管理 | [cb2eea6](http://47.99.38.141:10080/web/vango-erpweb/commits/cb2eea6)
 - | 12121212 | [e5a7068](http://47.99.38.141:10080/web/vango-erpweb/commits/e5a7068)
 - | 编写采购入库页面 | [7c1cb1e](http://47.99.38.141:10080/web/vango-erpweb/commits/7c1cb1e)
 - | 编写菜单管理 | [8b5e448](http://47.99.38.141:10080/web/vango-erpweb/commits/8b5e448)
 - | 编写菜单管理页面弹框 | [52b5f2a](http://47.99.38.141:10080/web/vango-erpweb/commits/52b5f2a)
 - | 编写查看、质检弹框 | [be3082e](http://47.99.38.141:10080/web/vango-erpweb/commits/be3082e)
 - | 编写订单页面 | [7d39d7b](http://47.99.38.141:10080/web/vango-erpweb/commits/7d39d7b)
 - | 菜单编辑 | [195a490](http://47.99.38.141:10080/web/vango-erpweb/commits/195a490)
 - | 菜单管理 | [65ad52c](http://47.99.38.141:10080/web/vango-erpweb/commits/65ad52c)
 侧边栏菜单 | 侧边栏菜单的完善 | [7092116](http://47.99.38.141:10080/web/vango-erpweb/commits/7092116)
 - | 导入订货管理上传文件调整 | [a0b64e2](http://47.99.38.141:10080/web/vango-erpweb/commits/a0b64e2)
 - | 调整导入订货单异常流程限制 | [c5674f7](http://47.99.38.141:10080/web/vango-erpweb/commits/c5674f7)
 - | 调整目录结构和文件名称 | [be36346](http://47.99.38.141:10080/web/vango-erpweb/commits/be36346)
 - | 调整下载文件的方法，订货管理默认查询近15天的内容 | [cc89292](http://47.99.38.141:10080/web/vango-erpweb/commits/cc89292)
 - | 对接接口 | [84a77e8](http://47.99.38.141:10080/web/vango-erpweb/commits/84a77e8)
 - | 汇率管理历史记录 没有结束时间的使用当前时间 | [e69b7fb](http://47.99.38.141:10080/web/vango-erpweb/commits/e69b7fb)
 - | 开发采购入库页面 | [db0e19e](http://47.99.38.141:10080/web/vango-erpweb/commits/db0e19e)
 - | 开发阶段侧边栏菜单渲染的完成 | [056b431](http://47.99.38.141:10080/web/vango-erpweb/commits/056b431)
 - | 联调后端接口，完善订货管理页面相关功能 | [962a016](http://47.99.38.141:10080/web/vango-erpweb/commits/962a016)
 路由新增了一个父级路由的template | 路由新增了一个父级路由的template | [afb542d](http://47.99.38.141:10080/web/vango-erpweb/commits/afb542d)
 - | 升级vue后遗症，调整方法引入方式 | [a2223df](http://47.99.38.141:10080/web/vango-erpweb/commits/a2223df)
 - | 提交菜单 | [f9e16f6](http://47.99.38.141:10080/web/vango-erpweb/commits/f9e16f6)
 - | 提交屏幕适配 | [28f0a8b](http://47.99.38.141:10080/web/vango-erpweb/commits/28f0a8b)
 - | 提交自定义组件 | [5692b87](http://47.99.38.141:10080/web/vango-erpweb/commits/5692b87)
 - | 提交elTable公用组件 | [36e623a](http://47.99.38.141:10080/web/vango-erpweb/commits/36e623a)
 - | 添加采购管理菜单、路由及文件，调整页面布局和css模块化，布局方案有待继续调整 | [2ebcd0d](http://47.99.38.141:10080/web/vango-erpweb/commits/2ebcd0d)
 - | 完善汇率管理接口请求及页面展示筛选等 | [0967dfb](http://47.99.38.141:10080/web/vango-erpweb/commits/0967dfb)
 - | 新增财务管理路由页面name | [95da9c9](http://47.99.38.141:10080/web/vango-erpweb/commits/95da9c9)
 - | 新增进度条 | [56f2c04](http://47.99.38.141:10080/web/vango-erpweb/commits/56f2c04)
 - | 新增路由历史记录模块 | [fd79ed3](http://47.99.38.141:10080/web/vango-erpweb/commits/fd79ed3)
 - | 新增刷新当前路由 | [811f4a8](http://47.99.38.141:10080/web/vango-erpweb/commits/811f4a8)
 - | 新增login.vue | [b77cc41](http://47.99.38.141:10080/web/vango-erpweb/commits/b77cc41)
 - | 修改菜单管理页面的复选框问题 | [0247457](http://47.99.38.141:10080/web/vango-erpweb/commits/0247457)
 - | 修改页面布局，并新增采购入库页面 | [0fdbcb3](http://47.99.38.141:10080/web/vango-erpweb/commits/0fdbcb3)
 - | 修改页面样式 | [40f1137](http://47.99.38.141:10080/web/vango-erpweb/commits/40f1137)
 - | 增加一个页面 | [3b33bad](http://47.99.38.141:10080/web/vango-erpweb/commits/3b33bad)


### 🎨 代码样式
范围|描述|commitId
--|--|--
 - | 写死token供开发 | [251fbdf](http://47.99.38.141:10080/web/vango-erpweb/commits/251fbdf)
 新增开发文件夹 | 新增配置的开发文件夹 | [db72dd8](http://47.99.38.141:10080/web/vango-erpweb/commits/db72dd8)
 新增开发文件夹 | 新增配置文件夹 | [bba56b4](http://47.99.38.141:10080/web/vango-erpweb/commits/bba56b4)


### 🐛 Bug 修复
范围|描述|commitId
--|--|--
 - | 创建代理的方法报错 | [3bee33f](http://47.99.38.141:10080/web/vango-erpweb/commits/3bee33f)

src/utils/tool.js
/*
 * @Descripttion: Vango ERP
 * @version: 1.0
 * @Author: 朱涛（梅溪）
 * @Date: 2021-07-06 12:54:54
 * @LastEditors: 朱涛
 * @LastEditTime: 2021-08-26 09:31:16
 */
import { ElMessage, ElMessageBox } from 'element-plus'
import  moment from 'moment'
import { v } from 'vxe-table'
export class TableTools {
    constructor(props) {
        if (props) this.refs = props
    }
    selectTableAll(data, flag) {
        for (let i = 0; i < data.length; i++) {
            this.refs.toggleRowSelection(data[i], flag)
            if (data[i].children != undefined) {
                this.selectTableAll(data[i].children, flag)
            }
        }
    }

    delItem(data, value, itemName) {
        for (let i = 0; i < data.length; i++) {
            let item = data[i][itemName]
            // 获取用来校验的值 ,itemName 是键值名
            if (item === value) {
                data.splice(i, 1)
                return false
            }
            if (data[i].children !== undefined && data[i].length > 0) {
                data[i].children = this.delItem(data[i].children, itemName, value)
            }
        }
        return data
    } // 递归删除table某一项
}
export class createFileName {
    constructor(props) {
        const { moduleName, type } = props
        this.moduleName = moduleName // 模块名
        this.type = type
    }
    createImportFileName = () => {
        return `${this.type}/${this.moduleName}/${this.createDate()}/${this.createTime()}_${this.moduleName}_${this.createMathNums(0, 10000)}`
    } // 生成导入文件的文件名

    createMathNums = (min, max) => {
        return Math.floor(Math.random() * (max - min + 1) + min)
    } //生成随机数
    createTime = () => {
        var date = new Date()
        let hour = String(date.getHours()).length == 2 ? String(date.getHours()) : ('0' + String(date.getHours()))
        let minu = String(date.getMinutes()).length == 2 ? String(date.getMinutes()) : ('0' + String(date.getMinutes()))
        let sec = String(date.getSeconds()).length == 2 ? String(date.getSeconds()) : ('0' + String(date.getSeconds()))
        return hour + minu + sec
    } // 获取时分秒
    createDate = () => {
        var date = new Date()
        let yy = String(date.getFullYear())
        let mm = String(date.getMonth()).length == 2 ? String(date.getMonth() + 1) : ( date.getMonth() + 1 === 10 ? String(date.getMonth() + 1) : '0' + String(date.getMonth() + 1))
        let dd = String(date.getDate()).length == 2 ? String(date.getDate()) : ('0' + String(date.getDate()))
        return yy + mm + dd
    } // 获取日期
}
export function type(para) {
    return Object.prototype.toString.call(para)
}

export function resetForm(form, filterForm) {
    for (var i in form) {
        if (type(form[i]) == '[object Array]' || type(form[i]) == 'Array') {
            form[i] = new Array()
        } else {
            form[i] = null
        }
        if (filterForm) form[i] = filterArr(i, form[i])
    }

    function filterArr(index, item) {
        let keyArr = Object.keys(filterForm)
        if (keyArr.includes(index)) {
            return filterForm[index]
        } else {
            return item
        }
    } // 如果有需要保留的值即可以同样的键值名 加需要保留的值 以对象方式传进来替换
    return form
} // 清空对象

export function showPermission(orderTabType, testOrderType, per) {
    let bole = true
    if (per) { } //权限校验模块
    if (orderTabType && testOrderType) {
        if (orderTabType !== testOrderType) bole = false
    }

    return bole
} // 根据权限和orderTabType 来决定是否显示

// 数组排序 Author: 王新国 (支持字符串排序，以编码顺序排序)
export function sortFn(arr, sortKey = 'orderNum') {
    if (arr.length == 0) return arr
    if (arr.filter(v => v[sortKey])) {
        arr.sort(function (a, b) {
            return a[sortKey] - b[sortKey]
        })
    } else {
        arr = arr.sort()
    }
    return arr
}

// 将list转为树状结构 Author: 王新国
//   function toTree(data, key, sortKey) {
//     key = key || 'deptId'
//     sortKey = sortKey || key
//     data = sortFn(data,sortKey);
export function toTree(data, key, parentId = 0) {
    key = key || 'deptId'
    sortFn(data)
    let result = []
    if (!Array.isArray(data)) {
        return result
    }
    if (data.length == 1) {
        return (result = data)
    }
    data.forEach(item => {
        // console.log(item,item.parentId)
        if (item.parentId === parentId) {
            // 递归寻找

            item.children = toTree(data, key, item[key])
            result.push(item)
        }
    })

    return result
}

/* 将树状结构转为list数组 Author: 王新国
 * 使用方法 [...treeToArr(treeArr,'children',true)]
 * 参数1：treeArr 需要转换的 树状结构数据 数据类型为Array       ==>必传项
 * 参数2：children 子集键名 数据类型为String                   ==>可传可不传，默认为children为子集键名
 * 参数3：完成转换后是否删除children子集 数据类型为Boolean      ==>可传可不传，默认为false，默认不删除子集
 */
export function* treeToArr(treeArr, children, deleteChildren) {
    children = children || 'children'
    deleteChildren = deleteChildren || false
    for (let item of treeArr) {
        let newItem = deepClone(item)
        delete newItem[children]
        yield newItem
        if (item[children] && item[children].length)
            yield* treeToArr(item[children])
    }
}

//将数组转树状结构，最后一层为list键名，特定场合使用 Author: 王新国
export function childrenToList(treeArr, key) {
    key = key || 'deptId'
    let result = []
    if (!Array.isArray(treeArr)) {
        return result
    }
    let map = {}
    treeArr.forEach(item => {
        if (item.menuType != 'F' && item.menuType != 'D') {
            delete item.children
            map[item[key]] = item
            let parent = map[item.parentId]
            if (parent) {
                ;
                (parent.children || (parent.children = [])).push(item)
            } else {
                result.push(item)
            }
        } else if (item.menuType == 'D') {
            delete item.children
            delete item.textList
            map[item[key]] = item
            let parent = map[item.parentId]
            if (parent) {
                ;
                (parent.textList || (parent.textList = [])).push(item)
            } else {
                result.push(item)
            }
        } else {
            delete item.children
            delete item.list
            map[item[key]] = item
            let parent = map[item.parentId]
            if (parent) {
                ;
                (parent.list || (parent.list = [])).push(item)
            } else {
                result.push(item)
            }
        }
    })
    return result
}

// 深拷贝 支持多层级数据深拷贝 Author: 王新国  注意：数组不可用此方法（主要使用deepCopy方法，deepClone方法用于多层对象的深拷贝）
export function deepClone(x) {
    const returnData = {}
    const loopList = [{
        parent: returnData,
        key: undefined,
        data: x
    }]
    while (loopList.length) {
        const node = loopList.pop()
        const parent = node.parent
        const key = node.key
        const data = node.data
        let res = parent
        if (typeof key !== 'undefined') {
            res = parent[key] = {}
        }
        for (let k in data) {
            if (data.hasOwnProperty(k)) {
                if (typeof data[k] === 'object') {
                    loopList.push({
                        parent: res,
                        key: k,
                        data: data[k]
                    })
                } else {
                    res[k] = data[k]
                }
            }
        }
    }
    return returnData
}

// 深拷贝 支持多层级数据深拷贝 Author: 王新国 通用，数组和对象都可以用
export function deepCopy(obj) {
    var result = Array.isArray(obj) ? [] : {}
    for (var key in obj) {
        if (obj.hasOwnProperty(key)) {
            if (typeof obj[key] === 'object' && obj[key] !== null) {
                result[key] = deepCopy(obj[key]) // 递归复制
            } else {
                result[key] = obj[key]
            }
        }
    }
    return result
}

// 禁止页面后退 Author: 韦剑
export function notAllowRouter() {
    history.pushState(null, null, document.URL);
    window.addEventListener('popstate', function () {
        history.pushState(null, null, document.URL);
    })
}

/*正则校验  Author: 王新国
 * 第一个参数 dataStr ==> 需要校验的字符串   类型：字符串
 * 第二个参数 type ==> 校验的分类   类型：字符串    可选值：'email'=>邮箱  'userName'=>用户名  'mobileNumber'=>手机号码 'roleName'=>角色名
 */
export function patternFn(dataStr, type) {
    let patternData = {
        email: /^\w+@\w+(\.\w+)+$/, //邮箱
        userName: /^[\s\u4e00-\u9fa5_a-zA-Z0-9_-]{2,16}$/, //用户名(中国人用)
        mobileNumber: /^((1[3,4,7,6,9,5,8,9][0-9]))\d{8}$/, //手机号码(国内手机号)
        roleName: /^[\s\u4e00-\u9fa5_a-zA-Z0-9_-]{2,16}$/, //角色名
        shinpping: /^[A-Za-z0-9]+$/, //运单号(数字+字母)
        postCode: /[\sa-zA-Z0-9]{1,40}/, //邮编（文本，数字、大写英文字母，最大40个字）
        taxNumber: /[\sa-zA-Z0-9]{1,30}/, //税号数字+大小写字母，最大30个字
        name: /[\sa-zA-Z]{1,60}/, //英文姓名(最大60个字)（外国人用）
        phone: /[0-9 -+()#]{1,60}$/, //国外电话：包含数字、+ - ( ) # 空格，最大60个字
        weght: /^[0-9]+(.[0-9]{0,5})?$/, //重量正则，5位有效小数
        webUrl: /^((http|ftp|https):\/\/[\w\-_]+(\.[\w\-_]+)+([\w\-\.,@?^=%&:/~\+#]*[\w\-\@?^=%&/~\+#])?)/,//网址正则
    }
    if (patternData[type].test(dataStr)) {
        return true
    } else {
        return false
    }
}

/*播放音频文件封装 Author: 王新国
 *第一个参数 audioUrl 音频文件路径 字符串
 *第二个参数 timeStamp 播放时长 毫秒数 默认3000毫秒
 */
export function audioPlay(audioUrl, timeStamp) {
    var body = document.getElementsByTagName("body")[0];
    var audio = document.createElement("audio");
    audio.setAttribute("id", "awardAudio");
    audio.setAttribute("autoplay", "autoplay");
    audio.setAttribute("src", audioUrl);
    body.appendChild(audio);
    setTimeout(function () {
        body.removeChild(audio);
    }, timeStamp);
}

//将字符串中的字标记为红色 Author: 王新国
export function reRed(str,arr){
	arr.forEach(v => {
		str = str.replace(v,'<span style="color:#F56C6C">' + v + '</span>')
	});
	return str
}


//打印proxy数据，方便调试(上线可去除) Author: 王新国
export function print(obj, str) {
    return console.log(str || '', JSON.parse(JSON.stringify(obj)))
}

// 防抖 Author: 韦剑
export function _debounce(fn, time = 500) {
    let timer = null
    return function (...args) {
        let that = this
        if (timer) clearTimeout(timer)
        timer = setTimeout(() => {
            fn.call(that, ...args)
        }, time);
    }
}


export function debounce(callBack) {
    let timer = null
    return function () {
        if (timer) clearTimeout(timer)
				let callNow = !timer
				timer = setTimeout(() => {
					timer = null
				}, 1000)
			if (callNow) callBack()
    };
}
// 防抖函数 Author: 朱涛


export function duplicate(array) {
    let obj = {}
    return array.filter((item, index) => {
        // 防止key重复
        let newItem = item + JSON.stringify(item)
        return obj.hasOwnProperty(newItem) ? false : obj[newItem] = true
    })

} // js 二维数组去重


export function confirm(text, callback,cancelCall) {
	ElMessageBox.confirm(text, '提示', {
		closeOnClickModal : false,
		confirmButtonText: '确定',
		cancelButtonText: '取消',
		type: 'warning'
	})
		.then(res => {
			callback()
		})
		.catch(err => {
			cancelCall()
		})
} //封装的提示弹窗

export function changeNull(val) {
    for (var i in val) {
        if (val[i] === '') val[i] = null
    }
    return val

}


export function copy(val){
    const textarea = document.createElement('textarea')
    textarea.readOnly = 'readonly'
    textarea.style.position = 'absolute'
    textarea.style.left = '-9999px'
    textarea.value = val
    document.body.appendChild(textarea)
    textarea.select()
    const result = document.execCommand('Copy')
    if (result) {
        ElMessage.success('复制成功') // 可根据项目UI仔细设计
    }
    document.body.removeChild(textarea)
}

 export function isObjectValueEqual(o1, o2) {
    //两个对象是否一样
    var props1 = Object.getOwnPropertyNames(o1)
    var props2 = Object.getOwnPropertyNames(o2)
    if (props1.length != props2.length) {
      return false
    }
    for (var i = 0, max = props1.length; i < max; i++) {
      var propName = props1[i]
      if (JSON.stringify(o1[propName]) !== JSON.stringify(o2[propName])) {
        console.log('内容比较有误==>',o1[propName],o2[propName])
        return false
      }
    }
    return true
  }

  export function findPathbyId(tree, id, path) {
      if (typeof path == 'undefined') {
          path = []
      }
      for (let i = 0; i < tree.length; i++) {
          let tempPath = [...path]
          tempPath.push(tree[i].id)
          if (tree[i].id === id) {
              return tempPath
          }
          if (tree[i].child) {
              let reuslt = findPathbyId(tree[i].child, id, tempPath)
              if (reuslt) {
                  return reuslt
              }
          }
      }
  }
  //  生成随机索引唯一标识 Author: 韦剑
export function getuuid() {
  var s = []
  var hexDigits = '0123456789abcdef'
  for (var i = 0; i < 36; i++) {
    s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1)
  }
  s[14] = '4' // bits 12-15 of the time_hi_and_version field to 0010
  s[19] = hexDigits.substr((s[19] & 0x3) | 0x8, 1) // bits 6-7 of the clock_seq_hi_and_reserved to 01
  s[8] = s[13] = s[18] = s[23] = '-'
  var uuid = s.join('')
  return uuid
}

// 场景：使用Object.assign时目标对象会被赋值某些自己没有而源对象有的属性
// usefulKeys 需要保留的源对象的属性值
export function assignTool(target, source, usefulKeys = [], insertObj = {}) {
  Object.keys(target).forEach(key => target[key] = source[key])
  usefulKeys.forEach(k => target[k] = source[k])
  Object.keys(insertObj).forEach(y => target[y] = insertObj[y])
  return target
}

// 很多SKU后端给json字符串，此函数解析返回拼接好可直接展示
export function skuShow(jsonStr){
    const jsonData = JSON.parse(jsonStr)
    if(jsonData){
        let str = ''
        jsonData.forEach(v=>{
            str += `${v.spe || v.speEn}:${v.speVal || v.speValEn} `
        })
        return str
    }
}
export function createId() {
	let timestamp = new Date().getTime()
	let num = parseInt((Math.random()*9+1) * 10000)
	// console.log(num)
	//console.log(timestamp+String(num))
	return timestamp + num
} // 生成随机id

export function printFail(cb){
	ElMessageBox.alert(
		'<p>请务必确认打印过程中是否出现以下情况？确定打印失败包裹会继续停留在当前状态以便再次打印！</p><p style="text-indent: 13px">1.pdf预览页面未弹出</p><p style="text-indent: 13px">2.pdf预览弹出但内容异常</p><p style="text-indent: 13px">3.在pdf预览页面点击打印后打印机未能正常打印或打印过程中出现异常</p><p style="font-weight: bold">若正常弹出pdf预览且打印机已正常打印请点击取消返回弹框点击打印完成，否者会出现面单多打的情况！</p>',
		'打印失败告警确认',
		{
			customStyle:'width:auto',
			type:'warning',
			dangerouslyUseHTMLString: true,
			showCancelButton	: true,
			confirmButtonText: '确定',
			cancelButtonText: '取消',
		}
	).then(()=>{
		cb()
	})

} // 打印面单二次确认弹框

// 格式化定制属性
export function renderCustomInfo(list = [], type = 1) {
	if (type == 1) {
		let str = ''
		list.forEach(v => str += `${v.field}:${v.value}`)
		return str
	} else {
		let arr = []
		list.forEach(v => {
			arr.push(v)
			if (v.subOptions) arr = [...arr, ...v.subOptions]
		})
		return arr
	}
}

// 是否是属于图片链接
export function isImgString(str = "") {
	const imgTypeList = ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.jfif']
	return imgTypeList.some(item=> str.indexOf(item) > -1 && (str.indexOf('https://') > -1 || str.indexOf('http://') > -1))
}


let formatStart = 'YYYY-MM-DD 00:00:00',formatEnd = 'YYYY-MM-DD',formatNow = 'YYYY-MM-DD HH:mm:ss'
const orderIndexParmas = {
		1: [moment().format(formatStart),moment().format(formatEnd)],//当前日期
		2: [moment().endOf('day').subtract(1, 'days').format(formatStart),moment().endOf('day').subtract(1, 'days').format(formatEnd)],//昨天日期
    3: [moment().endOf('day').subtract(3, 'days').format(formatStart),moment().endOf('day').subtract(1, 'days').format(formatEnd)],//三日日期
    4: [moment().endOf('day').subtract(7, 'days').format(formatStart),moment().endOf('day').subtract(1, 'days').format(formatEnd)],//近七天
    7: function(){
			let start  = moment().day(-1).format(formatStart)
			let end  = moment().format(formatNow)
			const days = moment().format('E')
			console.log(days)
			if(['0','6'].includes(days)){
				start = moment().day(6).format(formatStart)
				console.log(start)
			}
			return [start,end]
		}, // 本周日期
}
export {orderIndexParmas}
